### DISCUSSION TITLE
Sharing types between executable and libraries

### DISCUSSION THREAD

**Discussion by Tim---** [NationalSecurityAgency/ghidra#2675] [Q&A]
Hi,

I am trying to analyze an executable which depends on several custom libraries. 
To share the common structures and function definitions, I created a new project archive. So far so good.

It works, but having to commit/update for every single change is really painful. I often have to switch between the main program and the libraries to update the structure fields and function signatures and this means commit/update/conflict for each change.

Like any SCM, I understand that this overhead is needed, especially when several people are working on the same project. But for my need I am a single user, and the libraries are only used by a single program.
I would like to be able to work as if it was a single executable (synchronized function signatures, no structures duplication, etc...).
I wanted to use "Add to program" to add the libraries, but it only works for raw binaries.

So how can I avoid the extra versioning overhead if I don't need it ?
Would it be interesting (and even feasible) to import an ELF library to an ELF program the way rtld does ?

Also, working with the datatype archive raised some side questions:
1. when I rename the field of a structure on the decompiled view, the structure is not marked as "dirty". Should I open a bug ?
2. the only way I found to add a structure to a shared archive is to drap-and-drop it from the program archive to the shared archive. So I have to scroll through ALL the datatypes in the program archive to reach it. Is there a better way ? Would it be relevant to add a right click > Add to archive > "name of the archive" ?
3. if I update the signature of a library, I need to "Capture function Data Types" in the library, replace the original "function_name" signature in the shared archive with the "function_name.conflict", then "Apply function data types" in the main program. Is there a better way ? (especially to override the existing signature instead of creating a conflict, similar to how structures are handled).

Regards.

**Comment by gtackett** [comment]
I've already given this a thumbs-up but wanted to chime in that a "single user" mode for versioning would be _really_ nice.

I know that the official Ghidra folks have long recognized that Ghidra's current approach to data type management is really weak. Let's cross our fingers that they have something really amazing in the works. (I have no inside knowledge to that effect. 😄 )
